### DISCUSSION TITLE
The elegant way to install Brownie on Windows without encountering Cytoolz errors with venv working.

### DISCUSSION THREAD

**Discussion by n4n0b1t3** [smartcontractkit/full-blockchain-solidity-course-py#971] [General]
Edit: When using the latest Python version you might run into problems at the end of lesson 6.
If your issues are similar to this: [StackOverflow issue](https://stackoverflow.com/questions/71126128/brownie-testing-for-reverted-transactions-does-not-work-with-pytest-raises-or)
Please consider downgrading your Python to 3.9.10. But don't worry, you won't need to do this for your standard python installation, just for your project folder. Here is an article on how to do this: https://github.com/smartcontractkit/full-blockchain-solidity-course-py/discussions/1109

Please don't skip this article though, it contains an important part about the Visual C++ build tool installation.

If you are running into difficulties installing brownie as described in the [video in lesson 5](https://youtu.be/M576WGiDBdQ?t=16179), this article here might help to solve the issue. After the kind feedback from @cromewar that `pipx` is not necessary, I was able to simplify the installation significantly. AND!!! We can use Python's virtual environment. \o/ Here is the result:

Install part of the Visual C++ build tools: Download and install https://visualstudio.microsoft.com/visual-cpp-build-tools/
![image](https://user-images.githubusercontent.com/32838961/152692453-7061bfe3-56bd-41e7-abf0-6bd9a0c815da.png)

Select only the module ticked off in the screenshot (German system) and continue with the installation.

Open Powershell and create your project directory with e.g. 
`>mkdir brownie_simple_storage`
Change to the project directory
`>cd brownie_simple_storage`

Create a virtual environment
`brownie_simple_storage>python -m venv .venv`

Activate the virtual environment
`brownie_simple_storage>.\.venv\Scripts\Activate.ps1`

Optional: update pip
`(.venv) brownie_simple_storage>python.exe -m pip install --upgrade pip`

Important, otherwise you will get the infamous error: Install Cython
`(.venv) brownie_simple_storage>pip install Cython`

Finally, install brownie
`(.venv) brownie_simple_storage>pip install eth-brownie`

Test Brownie with
`(.venv) brownie_simple_storage>brownie --version`

Now... Brownie is a little shy and wants his own room. No open office for Brownie.
```
(.venv) brownie_simple_storage>mkdir sources
(.venv) brownie_simple_storage>cd sources
(.venv) sources>brownie init
```

OK, not finished yet. Brownie depends on "ganache-cli" which is deprecated but the new "ganache" works just fine. There is one step you need elevated rights, so you need to open Powershell as admin. (press Win key >enter "powershell" > right click on Powershell in start menu > select "run as administrator" > confirm > navigate to your project directory).

Activate your virtual environment
`brownie_simple_storage>.\.venv\Scripts\Activate.ps1`

Install Nodeenv
`(.venv) brownie_simple_storage>pip install nodeenv`

Connect nodeenv with your virtual Python environment  
```
(.venv) brownie_simple_storage>nodeenv -p
 * Install prebuilt node (17.4.0) ..... done.
```
Install ganache
`(.venv) brownie_simple_storage>npm install -g ganache`

Test ganache
`(.venv) brownie_simple_storage>ganache --version`

run *.py script with Brownie
```
(.venv) brownie_simple_storage>cd sources
(.venv) sources>brownie run .\scripts\deploy.py
```

Das Leben ist kein Ponnyhof!

**Comment by cromewar** [comment]
Nice one, as I already said have you consider writing articles of this?

**Reply by n4n0b1t3** [reply]
I am at the beginning of this journey. Maybe, when I have finished lesson 5, I will. 
**Question**: When running `Brownie compile`, do you have to be in your `contracts` directory? When I run this in the project root (one above the contracts dir) the compiler finishes without error but does not produce any files.

Patrick actually did this here in the video [4:32:56](https://youtu.be/M576WGiDBdQ?t=16376) for me this doesn't work. However, when I run `brownie compile` the contracts directory everything works fine, and the build>contracts>JSON file is produced.

This is baffeling.

**Reply by cromewar** [reply]
That's weird, as you should use the compile command on the root folder of the project, could you try again but first delete the build folder?

**Reply by n4n0b1t3** [reply]
super weird, I just removed the JSON file, tried to recreate the issue and now it works. 
About articles at Medium. Which of my posts do you deem worthy for Medium?

**Reply by n4n0b1t3** [reply]
like this: https://n4n0b1t3.medium.com/running-brownie-on-windows-in-a-virtual-environment-without-encountering-cytoolz-errors-d0095fad905

**Reply by cromewar** [reply]
> super weird, I just removed the JSON file, tried to recreate the issue and now it works. About articles at Medium. Which of my posts do you deem worthy for Medium?

Interesting, not sure why it happened tough.

**Reply by cromewar** [reply]
> like this: https://n4n0b1t3.medium.com/running-brownie-on-windows-in-a-virtual-environment-without-encountering-cytoolz-errors-d0095fad905

Nice!!

I would say you can create a compilation of the tips you gave us and create like a one big article.. may be?
